File: 이분탐색/boj7453_lsh.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
#include <algorithm>

using namespace std;
using ll = long long;

int main(void)
{
  int N, num, left, right;
  int A[4001];
  int B[4001];
  int C[4001];
  int D[4001];
  vector<ll> AB, CD;

  cin >> N;

  for (int n = 0; N > n; n++)
    scanf("%d %d %d %d", &A[n], &B[n], &C[n], &D[n]);

  for (int i = 0; N > i; i++)
    for (int j = 0; N > j; j++)
    {
      AB.push_back(A[i] + B[j]);
      CD.push_back(C[i] + D[j]);
    }

  sort(AB.begin(), AB.end());
  sort(CD.begin(), CD.end());

  ll hi = AB.size(), length = AB.size(), cnt = 0;

  for (int i = 0; length > i; i++)
  {
    num = AB[i];
    left = lower_bound(CD.begin(), CD.begin() + hi, -num) - CD.begin();
    right = upper_bound(CD.begin(), CD.begin() + hi, -num) - CD.begin();

    if (left == hi || CD[left] != -num)
      continue;
    cnt += (right - left);
    hi = right;
  }

  cout << cnt << '\n';

  return 0;
}

File: 이분탐색/boj7453_lsh.py
import bisect

n = int(input())

a, b, c, d = [], [], [], []
for _ in range(n):
    q,w,e,r = map(int, input().split())
    a.append(q)
    b.append(w)
    c.append(e)
    d.append(r)


ab = []
cd = []

for i in range(n):
    for j in range(n):
        ab.append(a[i] + b[j])
        cd.append(c[i] + d[j])

cnt = 0
ab.sort()
cd.sort()

hi = len(ab)

for i in range(len(ab)):
    num = ab[i]
    left = bisect.bisect_left(cd, -num)
    right = bisect.bisect_right(cd, -num)

    if left == hi or cd[left] != -num:
        continue
    cnt += (right - left)
    hi = right
print(cnt)

